app/src/chat_api.py

```diff
@@ -0,0 +1,240 @@
+    # Set the LiteralAI user ID for this session so it can be used in literalai().thread()
+    session.literalai_user_id = literalai_user.id
+    return session
+
+
+# endregion
+# region: ===================  Example API Endpoint and logging to LiteralAI  ===================
+
+
+# Make sure to use async functions for faster responses
+@router.get("/engines")
+async def engines(user_id: str) -> list[str]:
+    session = await _get_user_session(user_id)
+    # Example of using Literal AI to log the request and response
+    with literalai().thread(name="API:/engines", participant_id=session.literalai_user_id):
+        request_msg = literalai().message(
+            content="List chat engines",
+            type="user_message",
+            name=user_id,
+            metadata=session.user.__dict__,
+        )
+        # This will show up as a separate step in LiteralAI, showing input and output
+        with literalai().step(type="tool"):
+            response = [
+                engine
+                for engine in chat_engine.available_engines()
+                if engine in session.user.allowed_engines
+            ]
+        # Example of using parent_id to have a hierarchy of messages in Literal AI
+        literalai().message(content=str(response), type="system_message", parent_id=request_msg.id)
+
+    return response
+
+
+# endregion
+# region: ===================  API Endpoints  ===================
+
+
+class QueryRequest(BaseModel):
+    session_id: str
+    new_session: bool
+    message: str
+
+    user_id: Optional[str] = None
+    agency_id: Optional[str] = None
+    beneficiary_id: Optional[str] = None
+
+
+class Citation(BaseModel):
+    citation_id: str
+    source_id: str
+    source_name: str
+    page_number: Optional[int] | None
+    uri: Optional[str] | None
+    headings: Optional[list[str]] | None
+    citation_text: str
+
+    @staticmethod
+    def from_subsection(subsection: Subsection) -> "Citation":
+        chunk = subsection.chunk
+        return Citation(
+            citation_id=f"citation-{subsection.id}",
+            source_id=str(chunk.document.id),
+            source_name=chunk.document.name,
+            page_number=chunk.page_number,
+            uri=chunk.document.source,
+            headings=chunk.headings,
+            citation_text=subsection.text,
+        )
+
+
+class QueryResponse(BaseModel):
+    response_text: str
+    citations: list[Citation]
+
+    # Populated after instantiation based on LiteralAI message ID
+    response_id: Optional[str] = None
+
+
+def get_chat_engine(session: UserSession) -> ChatEngineInterface:
+    engine_id = session.chat_engine_settings.engine_id
+    # May want to cache engine instances rather than creating them for each request
+    engine = (
+        chat_engine.create_engine(engine_id) if engine_id in session.user.allowed_engines else None
+    )
+    if not engine:
+        raise HTTPException(status_code=406, detail=f"Unknown engine: {engine_id}")
```

Comment:
strictly speaking I think this should be 401 (though TIL about 406!)

Reply:
401 sounds like an authentication problem, which this is not. How about 403?
> similar to [401](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/401), except that for 403 Forbidden responses, authenticating or re-authenticating makes no difference. The request failure is tied to application logic, such as insufficient permissions to a resource or action.